src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
+
+                    if (!loopBlocks.isMember(block->bbNum))
+                    {
+                        // This code is lexically between TOP and BOTTOM, but it does not
+                        // participate in the flow cycle.  Check for a run of consecutive
+                        // such blocks.
+                        BasicBlock* lastExitBlock = block;
```

Comment:
`lastExitBlock` sounds a lot like a name for a block that is in a loop. Maybe `lastNonLoopBlock`?